libsplinter/src/network/dispatch.rs

```diff
@@ -301,112 +295,144 @@ impl<MT: Hash + Eq + Debug + Clone> HandlerWrapper<MT> {
 
 /// A message to be dispatched.
 ///
-/// This struct contains information about a message that will be passed to a `Dispatcher` instance
+/// This enum contains information about a message that will be passed to a `Dispatcher` instance
 /// via a `Sender<DispatchMessage>`.
 #[derive(Clone)]
-pub struct DispatchMessage<MT: Any + Hash + Eq + Debug + Clone> {
-    message_type: MT,
-    message_bytes: Vec<u8>,
-    source_peer_id: String,
+enum DispatchMessage<MT: Any + Hash + Eq + Debug + Clone> {
+    Message {
+        message_type: MT,
+        message_bytes: Vec<u8>,
+        source_peer_id: String,
+    },
+    Shutdown,
 }
 
-impl<MT: Any + Hash + Eq + Debug + Clone> DispatchMessage<MT> {
-    /// Constructs a new DispatchMessage
-    pub fn new(message_type: MT, message_bytes: Vec<u8>, source_peer_id: String) -> Self {
-        DispatchMessage {
-            message_type,
-            message_bytes,
-            source_peer_id,
+pub struct DispatchLoopShutdownSignaler<MT: Any + Hash + Eq + Debug + Clone> {
+    sender: Sender<DispatchMessage<MT>>,
+}
+
+impl<MT: Any + Hash + Eq + Debug + Clone> DispatchLoopShutdownSignaler<MT> {
+    pub fn shutdown(&self) {
+        if self.sender.send(DispatchMessage::Shutdown).is_err() {
+            error!("Unable to send shutdown signal to already shutdown dispatch loop");
         }
     }
+}
 
-    pub fn message_type(&self) -> &MT {
-        &self.message_type
+/// Errors that may occur during the operation of the Dispatch Loop.
+#[derive(Debug)]
+pub struct DispatchLoopError(String);
+
+#[derive(Default)]
+pub struct DispatchLoopBuilder<MT: Any + Hash + Eq + Debug + Clone> {
+    dispatcher: Option<Dispatcher<MT>>,
+}
+
+impl<MT: Any + Hash + Eq + Debug + Clone + Send> DispatchLoopBuilder<MT> {
+    pub fn new() -> Self {
+        DispatchLoopBuilder { dispatcher: None }
     }
 
-    pub fn message_bytes(&self) -> &[u8] {
-        &self.message_bytes
+    pub fn with_dispatcher(mut self, dispatcher: Dispatcher<MT>) -> Self {
+        self.dispatcher = Some(dispatcher);
+        self
     }
 
-    pub fn source_peer_id(&self) -> &str {
-        &self.source_peer_id
+    pub fn build(mut self) -> Result<DispatchLoop<MT>, String> {
+        let (tx, rx) = channel();
+
+        let dispatcher = self
+            .dispatcher
+            .take()
+            .ok_or_else(|| "No dispatch provided".to_string())?;
+
+        let join_handle = std::thread::spawn(move || loop {
```

Comment:
use thread::Builder to give this thread a good name - https://doc.rust-lang.org/std/thread/#configuring-threads